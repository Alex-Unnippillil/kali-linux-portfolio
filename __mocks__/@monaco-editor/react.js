import React from 'react';
export default function MonacoMock({ value = '', onChange }) {
  return React.createElement('textarea', {
    value,
    onChange: e => onChange && onChange(e.target.value),
  });
}
