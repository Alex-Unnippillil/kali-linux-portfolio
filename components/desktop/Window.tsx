"use client";

import React, { useCallback, useEffect, useRef, useState } from "react";

interface WindowProps {
  title: string;
  children?: React.ReactNode;
  initialX?: number;
  initialY?: number;
  initialWidth?: number;
  initialHeight?: number;
  onClose?: () => void;
  onMinimize?: () => void;
  onMaximize?: () => void;
}

let zCounter = 1;

interface SnapPreview {
  left: number;
  top: number;
  width: number;
  height: number;
}

type SnapPosition = "left" | "right" | "max" | null;

const EDGE_THRESHOLD = 30;

const Window: React.FC<WindowProps> = ({
  title,
  children,
  initialX = 80,
  initialY = 80,
  initialWidth = 400,
  initialHeight = 300,
  onClose,
  onMinimize,
  onMaximize,
}) => {
  const winRef = useRef<HTMLDivElement>(null);
  const [pos, setPos] = useState({ x: initialX, y: initialY });
  const [size, setSize] = useState({ w: initialWidth, h: initialHeight });
  const [dragging, setDragging] = useState(false);
  const [resizeDir, setResizeDir] = useState<string | null>(null);
  const [snapPreview, setSnapPreview] = useState<SnapPreview | null>(null);
  const [snapPosition, setSnapPosition] = useState<SnapPosition>(null);
  const dragOffset = useRef({ x: 0, y: 0 });
  const [zIndex, setZIndex] = useState(++zCounter);
  const [focused, setFocused] = useState(false);
  const [prefersReducedMotion, setPrefersReducedMotion] = useState(false);

  useEffect(() => {
    const media = window.matchMedia("(prefers-reduced-motion: reduce)");
    const update = () => setPrefersReducedMotion(media.matches);
    update();
    media.addEventListener("change", update);
    return () => media.removeEventListener("change", update);
  }, []);

  const bringToFront = useCallback(() => {
    setZIndex(++zCounter);
    setFocused(true);
    document.dispatchEvent(new CustomEvent("desktop-window-focus", { detail: winRef.current }));
  }, []);

  useEffect(() => {
    const handler = (e: Event) => {
      const target = (e as CustomEvent).detail as HTMLDivElement;
      if (target !== winRef.current) setFocused(false);
    };
    document.addEventListener("desktop-window-focus", handler);
    return () => document.removeEventListener("desktop-window-focus", handler);
  }, []);

  const handleHeaderPointerDown = (e: React.PointerEvent) => {
    bringToFront();
    setDragging(true);
    dragOffset.current = { x: e.clientX - pos.x, y: e.clientY - pos.y };
    e.preventDefault();
  };

  const startResize = (dir: string) => (e: React.PointerEvent) => {
    bringToFront();
    setResizeDir(dir);
    e.stopPropagation();
    e.preventDefault();
  };

  const checkSnapPreview = (x: number, y: number) => {
    const left = x;
    const top = y;
    const right = x + size.w;
    if (left <= EDGE_THRESHOLD) {
      setSnapPreview({ left: 0, top: 0, width: window.innerWidth / 2, height: window.innerHeight });
      setSnapPosition("left");
    } else if (right >= window.innerWidth - EDGE_THRESHOLD) {
      setSnapPreview({ left: window.innerWidth / 2, top: 0, width: window.innerWidth / 2, height: window.innerHeight });
      setSnapPosition("right");
    } else if (top <= EDGE_THRESHOLD) {
      setSnapPreview({ left: 0, top: 0, width: window.innerWidth, height: window.innerHeight });
      setSnapPosition("max");
    } else if (snapPreview) {
      setSnapPreview(null);
      setSnapPosition(null);
    }
  };

  const snapWindow = useCallback(
    (pos: SnapPosition) => {
      if (!pos) return;
      if (pos === "left") {
        setPos({ x: 0, y: 0 });
        setSize({ w: window.innerWidth / 2, h: window.innerHeight });
      } else if (pos === "right") {
        setPos({ x: window.innerWidth / 2, y: 0 });
        setSize({ w: window.innerWidth / 2, h: window.innerHeight });
      } else if (pos === "max") {
        setPos({ x: 0, y: 0 });
        setSize({ w: window.innerWidth, h: window.innerHeight });
      }
    },
    []
  );

  useEffect(() => {
    const handleMove = (e: PointerEvent) => {
      if (dragging) {
        const newX = e.clientX - dragOffset.current.x;
        const newY = e.clientY - dragOffset.current.y;
        checkSnapPreview(newX, newY);
        setPos({ x: newX, y: newY });
      } else if (resizeDir) {
        setSize((prev) => {
          let { w, h } = prev;
          let { x, y } = pos;
          if (resizeDir.includes("right")) w = e.clientX - x;
          if (resizeDir.includes("bottom")) h = e.clientY - y;
          if (resizeDir.includes("left")) {
            const newX = e.clientX;
            w += x - newX;
            x = newX;
          }
          if (resizeDir.includes("top")) {
            const newY = e.clientY;
            h += y - newY;
            y = newY;
          }
          setPos({ x, y });
          return { w, h };
        });
      }
    };

    const handleUp = () => {
      if (dragging) snapWindow(snapPosition);
      setDragging(false);
      setResizeDir(null);
      setSnapPreview(null);
      setSnapPosition(null);
    };

    window.addEventListener("pointermove", handleMove);
    window.addEventListener("pointerup", handleUp);
    return () => {
      window.removeEventListener("pointermove", handleMove);
      window.removeEventListener("pointerup", handleUp);
    };
  }, [dragging, resizeDir, pos, snapPosition, snapWindow]);

  return (
    <>
      {snapPreview && (
        <div
          data-testid="snap-preview"
          className={`fixed z-50 pointer-events-none border-2 border-dashed border-accent bg-accent/20 rounded ${
            prefersReducedMotion ? "" : "transition-all duration-150"
          }`}
          style={snapPreview}
        />
      )}
      <div
        ref={winRef}
        onPointerDown={bringToFront}
        style={{
          top: pos.y,
          left: pos.x,
          width: size.w,
          height: size.h,
          zIndex,
          transition:
            dragging || resizeDir
              ? "none"
              : prefersReducedMotion
              ? "none"
              : "all 0.15s ease",
        }}
        tabIndex={0}
        className={`absolute bg-white shadow-lg rounded-lg border border-gray-300 overflow-hidden focus:outline-none focus-visible:ring-2 focus-visible:ring-accent focus-visible:ring-offset-2 rm:transition-none ${
          focused ? "" : "opacity-90"
        }`}
      >
      <div
        className="h-8 bg-gray-200 flex items-center cursor-move rounded-t-lg select-none"
        onPointerDown={handleHeaderPointerDown}
      >
        <div className="flex space-x-2 px-2">
          <button
            aria-label="Close"
            onClick={onClose}
            className="w-3 h-3 bg-red-500 rounded-full"
          />
          <button
            aria-label="Minimize"
            onClick={onMinimize}
            className="w-3 h-3 bg-yellow-500 rounded-full"
          />
          <button
            aria-label="Maximize"
            onClick={onMaximize}
            className="w-3 h-3 bg-green-500 rounded-full"
          />
        </div>
        <span className="flex-1 text-center pr-6 text-sm">{title}</span>
      </div>
      <div className="w-full h-[calc(100%-2rem)]">{children}</div>

      {/* Resize handles */}
      <div className="absolute top-0 left-0 w-full h-4 -mt-2 cursor-n-resize hover:bg-accent/20" onPointerDown={startResize("top")} />
      <div className="absolute bottom-0 left-0 w-full h-4 -mb-2 cursor-s-resize hover:bg-accent/20" onPointerDown={startResize("bottom")} />
      <div className="absolute top-0 left-0 h-full w-4 -ml-2 cursor-w-resize hover:bg-accent/20" onPointerDown={startResize("left")} />
      <div className="absolute top-0 right-0 h-full w-4 -mr-2 cursor-e-resize hover:bg-accent/20" onPointerDown={startResize("right")} />
      <div className="absolute top-0 left-0 w-4 h-4 -mt-2 -ml-2 cursor-nw-resize hover:bg-accent/20" onPointerDown={startResize("top-left")} />
      <div className="absolute top-0 right-0 w-4 h-4 -mt-2 -mr-2 cursor-ne-resize hover:bg-accent/20" onPointerDown={startResize("top-right")} />
        <div className="absolute bottom-0 left-0 w-4 h-4 -mb-2 -ml-2 cursor-sw-resize hover:bg-accent/20" onPointerDown={startResize("bottom-left")} />
        <div className="absolute bottom-0 right-0 w-4 h-4 -mb-2 -mr-2 cursor-se-resize hover:bg-accent/20" onPointerDown={startResize("bottom-right")} />
      </div>
    </>
  );
};

export default Window;

