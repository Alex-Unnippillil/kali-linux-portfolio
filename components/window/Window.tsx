import React, {
  useImperativeHandle,
  useRef,
  useState,
  forwardRef,
} from "react";
import Draggable, { DraggableEventHandler } from "react-draggable";

interface WindowProps {
  children: React.ReactNode;
}

interface SnapPreview {
  left: string;
  top: string;
  width: string;
  height: string;
}

type SnapPosition =
  | "left"
  | "right"
  | "top"
  | "bottom"
  | "top-left"
  | "top-right"
  | "bottom-left"
  | "bottom-right"
  | "left-third"
  | "center-third"
  | "right-third"
  | "max"
  | null;

const EDGE_THRESHOLD = 30;

const Window = forwardRef<{ handleDrag: () => void; handleStop: () => void; handleKeyDown: (e: any) => void; }, WindowProps>(
  ({ children }, ref) => {
    const nodeRef = useRef<HTMLDivElement>(null);
    const [snapPreview, setSnapPreview] = useState<SnapPreview | null>(null);
    const [snapPosition, setSnapPosition] = useState<SnapPosition>(null);
    const [snapped, setSnapped] = useState<SnapPosition>(null);
    const [width, setWidth] = useState(60);
    const [height, setHeight] = useState(85);
    const [dragging, setDragging] = useState(false);

    const checkSnapPreview: DraggableEventHandler = () => {
      const node = nodeRef.current;
      if (!node) return;
      const rect = node.getBoundingClientRect();
      const midX = rect.left + rect.width / 2;
      const thirdW = window.innerWidth / 3;

      if (rect.left <= EDGE_THRESHOLD && rect.top <= EDGE_THRESHOLD) {
        setSnapPreview({ left: "0", top: "0", width: "50%", height: "50%" });
        setSnapPosition("top-left");
      } else if (rect.right >= window.innerWidth - EDGE_THRESHOLD && rect.top <= EDGE_THRESHOLD) {
        setSnapPreview({ left: "50%", top: "0", width: "50%", height: "50%" });
        setSnapPosition("top-right");
      } else if (rect.left <= EDGE_THRESHOLD && rect.bottom >= window.innerHeight - EDGE_THRESHOLD) {
        setSnapPreview({ left: "0", top: "50%", width: "50%", height: "50%" });
        setSnapPosition("bottom-left");
      } else if (rect.right >= window.innerWidth - EDGE_THRESHOLD && rect.bottom >= window.innerHeight - EDGE_THRESHOLD) {
        setSnapPreview({ left: "50%", top: "50%", width: "50%", height: "50%" });
        setSnapPosition("bottom-right");
      } else if (rect.top <= EDGE_THRESHOLD) {
        setSnapPreview({ left: "0", top: "0", width: "100%", height: "50%" });
        setSnapPosition("top");
      } else if (rect.bottom >= window.innerHeight - EDGE_THRESHOLD) {
        setSnapPreview({ left: "0", top: "50%", width: "100%", height: "50%" });
        setSnapPosition("bottom");
      } else if (rect.left <= EDGE_THRESHOLD) {
        setSnapPreview({ left: "0", top: "0", width: "50%", height: "100%" });
        setSnapPosition("left");
      } else if (rect.right >= window.innerWidth - EDGE_THRESHOLD) {
        setSnapPreview({ left: "50%", top: "0", width: "50%", height: "100%" });
        setSnapPosition("right");
      } else if (midX < thirdW) {
        setSnapPreview({ left: "0", top: "0", width: "33.33%", height: "100%" });
        setSnapPosition("left-third");
      } else if (midX >= thirdW && midX < 2 * thirdW) {
        setSnapPreview({ left: "33.33%", top: "0", width: "33.33%", height: "100%" });
        setSnapPosition("center-third");
      } else if (midX >= 2 * thirdW) {
        setSnapPreview({ left: "66.66%", top: "0", width: "33.33%", height: "100%" });
        setSnapPosition("right-third");
      } else if (snapPreview) {
        setSnapPreview(null);
        setSnapPosition(null);
      }
    };

    const handleDrag: DraggableEventHandler = (...args) => {
      setDragging(true);
      checkSnapPreview(...args);
    };

    const snapWindow = (pos: SnapPosition) => {
      const node = nodeRef.current;
      if (!node || !pos) return;
      if (pos === "left") {
        node.style.transform = `translate(0px, 0px)`;
        setWidth(50);
        setHeight(100);
      } else if (pos === "right") {
        node.style.transform = `translate(${window.innerWidth / 2}px, 0px)`;
        setWidth(50);
        setHeight(100);
      } else if (pos === "top") {
        node.style.transform = `translate(0px, 0px)`;
        setWidth(100);
        setHeight(50);
      } else if (pos === "bottom") {
        node.style.transform = `translate(0px, ${window.innerHeight / 2}px)`;
        setWidth(100);
        setHeight(50);
      } else if (pos === "top-left") {
        node.style.transform = `translate(0px, 0px)`;
        setWidth(50);
        setHeight(50);
      } else if (pos === "top-right") {
        node.style.transform = `translate(${window.innerWidth / 2}px, 0px)`;
        setWidth(50);
        setHeight(50);
      } else if (pos === "bottom-left") {
        node.style.transform = `translate(0px, ${window.innerHeight / 2}px)`;
        setWidth(50);
        setHeight(50);
      } else if (pos === "bottom-right") {
        node.style.transform = `translate(${window.innerWidth / 2}px, ${window.innerHeight / 2}px)`;
        setWidth(50);
        setHeight(50);
      } else if (pos === "left-third") {
        node.style.transform = `translate(0px, 0px)`;
        setWidth(33.33);
        setHeight(100);
      } else if (pos === "center-third") {
        node.style.transform = `translate(${window.innerWidth / 3}px, 0px)`;
        setWidth(33.33);
        setHeight(100);
      } else if (pos === "right-third") {
        node.style.transform = `translate(${(2 * window.innerWidth) / 3}px, 0px)`;
        setWidth(33.33);
        setHeight(100);
      } else if (pos === "max") {
        node.style.transform = `translate(0px, 0px)`;
        setWidth(100);
        setHeight(96.3);
      }
      setSnapped(pos);
    };

    const handleStop: DraggableEventHandler = () => {
      snapWindow(snapPosition);
      setSnapPreview(null);
      setSnapPosition(null);
      setDragging(false);
    };

    const handleKeyDown = (e: any) => {
      if (snapPreview && e.key === "Escape") {
        e.preventDefault?.();
        e.stopPropagation?.();
        setSnapPreview(null);
        setSnapPosition(null);
        return;
      }
      if (e.altKey) {
        if (e.key === "ArrowLeft") {
          e.preventDefault?.();
          snapWindow("left");
        } else if (e.key === "ArrowRight") {
          e.preventDefault?.();
          snapWindow("right");
        } else if (e.key === "ArrowUp") {
          e.preventDefault?.();
          snapWindow("max");
        }
      }
    };

    useImperativeHandle(ref, () => ({ handleDrag, handleStop, handleKeyDown }));

    const guides: SnapPreview[] = [
      { left: "0", top: "0", width: "50%", height: "100%" },
      { left: "50%", top: "0", width: "50%", height: "100%" },
      { left: "0", top: "0", width: "100%", height: "50%" },
      { left: "0", top: "50%", width: "100%", height: "50%" },
      { left: "0", top: "0", width: "50%", height: "50%" },
      { left: "50%", top: "0", width: "50%", height: "50%" },
      { left: "0", top: "50%", width: "50%", height: "50%" },
      { left: "50%", top: "50%", width: "50%", height: "50%" },
      { left: "0", top: "0", width: "33.33%", height: "100%" },
      { left: "33.33%", top: "0", width: "33.33%", height: "100%" },
      { left: "66.66%", top: "0", width: "33.33%", height: "100%" },
    ];

    return (
      <>
        {dragging &&
          guides.map((g, i) => (
            <div
              key={i}
              className="fixed z-40 pointer-events-none border border-dashed border-white/20 bg-white/5"
              style={g}
            />
          ))}
        {snapPreview && (
          <div
            data-testid="snap-preview"
            className="fixed z-50 pointer-events-none border-2 border-dashed border-white bg-white bg-opacity-20"
            style={snapPreview}
          />
        )}
        <Draggable onStart={() => setDragging(true)} onDrag={handleDrag} onStop={handleStop} nodeRef={nodeRef}>
          <div
            ref={nodeRef}
            tabIndex={0}
            onKeyDown={handleKeyDown}
            style={{ width: `${width}%`, height: `${height}%` }}
          >
            {children}
          </div>
        </Draggable>
      </>
    );
  }
);

export default Window;
