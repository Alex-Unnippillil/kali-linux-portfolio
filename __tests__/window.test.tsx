import React, { act } from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import Window from '../components/base/window';

const setViewport = (width: number, height: number) => {
  Object.defineProperty(window, 'innerWidth', { configurable: true, writable: true, value: width });
  Object.defineProperty(window, 'innerHeight', { configurable: true, writable: true, value: height });
};

beforeEach(() => {
  setViewport(1440, 900);
});

jest.mock('react-ga4', () => ({ send: jest.fn(), event: jest.fn() }));
jest.mock('react-draggable', () => ({
  __esModule: true,
  default: ({ children }: { children: React.ReactNode }) => <div>{children}</div>,
}));
jest.mock('../components/apps/terminal', () => ({ displayTerminal: jest.fn() }));

describe('Window lifecycle', () => {
  it('invokes callbacks on close', () => {
    jest.useFakeTimers();
    const closed = jest.fn();
    const hideSideBar = jest.fn();

    render(
      <Window
        id="test-window"
        title="Test"
        screen={() => <div>content</div>}
        focus={() => {}}
        hasMinimised={() => {}}
        closed={closed}
        hideSideBar={hideSideBar}
        openApp={() => {}}
      />
    );

    const closeButton = screen.getByRole('button', { name: /window close/i });
    fireEvent.click(closeButton);

    expect(hideSideBar).toHaveBeenCalledWith('test-window', false);

    act(() => {
      jest.advanceTimersByTime(300);
    });

    expect(closed).toHaveBeenCalledWith('test-window');
    jest.useRealTimers();
  });
});

describe('Window snapping preview', () => {
  it('shows preview when dragged near left edge', () => {
    setViewport(1920, 1080);
    const ref = React.createRef<Window>();
    render(
      <Window
        id="test-window"
        title="Test"
        screen={() => <div>content</div>}
        focus={() => {}}
        hasMinimised={() => {}}
        closed={() => {}}
        hideSideBar={() => {}}
        openApp={() => {}}
        ref={ref}
      />
    );

    const winEl = document.getElementById('test-window')!;
    // Simulate being near the left edge
    winEl.getBoundingClientRect = () => ({
      left: 5,
      top: 150,
      right: 105,
      bottom: 250,
      width: 100,
      height: 100,
      x: 5,
      y: 150,
      toJSON: () => {}
    });

    act(() => {
      ref.current!.handleDrag();
    });

    const preview = screen.getByTestId('snap-preview');
    expect(preview).toBeInTheDocument();
    expect((preview as HTMLElement).style.backdropFilter).toBe('brightness(1.2)');
  });

  it('hides preview when away from edge', () => {
    const ref = React.createRef<Window>();
    render(
      <Window
        id="test-window"
        title="Test"
        screen={() => <div>content</div>}
        focus={() => {}}
        hasMinimised={() => {}}
        closed={() => {}}
        hideSideBar={() => {}}
        openApp={() => {}}
        ref={ref}
      />
    );

    const winEl = document.getElementById('test-window')!;
    // Position far from edges
    winEl.getBoundingClientRect = () => ({
      left: 200,
      top: 200,
      right: 300,
      bottom: 300,
      width: 100,
      height: 100,
      x: 200,
      y: 200,
      toJSON: () => {}
    });

    act(() => {
      ref.current!.handleDrag();
    });

    expect(screen.queryByTestId('snap-preview')).toBeNull();
  });

  it('shows top preview when dragged near top edge', () => {
    setViewport(1280, 720);
    const ref = React.createRef<Window>();
    render(
      <Window
        id="test-window"
        title="Test"
        screen={() => <div>content</div>}
        focus={() => {}}
        hasMinimised={() => {}}
        closed={() => {}}
        hideSideBar={() => {}}
        openApp={() => {}}
        ref={ref}
      />
    );

    const winEl = document.getElementById('test-window')!;
    winEl.getBoundingClientRect = () => ({
      left: 400,
      top: 5,
      right: 500,
      bottom: 105,
      width: 100,
      height: 100,
      x: 400,
      y: 5,
      toJSON: () => {}
    });

    act(() => {
      ref.current!.handleDrag();
    });

    expect(ref.current!.state.snapPosition).toBe('top');
    const preview = screen.getByTestId('snap-preview');
    expect(preview).toHaveStyle(`height: ${window.innerHeight / 2}px`);
  });
});

describe('Window snapping finalize and release', () => {
  it('snaps window on drag stop near left edge', () => {
    setViewport(1024, 768);
    const ref = React.createRef<Window>();
    render(
      <Window
        id="test-window"
        title="Test"
        screen={() => <div>content</div>}
        focus={() => {}}
        hasMinimised={() => {}}
        closed={() => {}}
        hideSideBar={() => {}}
        openApp={() => {}}
        ref={ref}
      />
    );

    const winEl = document.getElementById('test-window')!;
    winEl.getBoundingClientRect = () => ({
      left: 5,
      top: 120,
      right: 105,
      bottom: 220,
      width: 100,
      height: 100,
      x: 5,
      y: 120,
      toJSON: () => {}
    });

    act(() => {
      ref.current!.handleDrag();
    });
    act(() => {
      ref.current!.handleStop();
    });

    expect(ref.current!.state.snapped).toBe('left');
    expect(ref.current!.state.width).toBeCloseTo(50);
    const expectedHeight = ((window.innerHeight - 28) / window.innerHeight) * 100;
    expect(ref.current!.state.height).toBeCloseTo(expectedHeight, 5);
    expect(winEl.style.transform).toBe('translate(0px, 0px)');
  });

  it('snaps window on drag stop near right edge on large viewport', () => {
    setViewport(1920, 1080);
    const ref = React.createRef<Window>();
    render(
      <Window
        id="test-window"
        title="Test"
        screen={() => <div>content</div>}
        focus={() => {}}
        hasMinimised={() => {}}
        closed={() => {}}
        hideSideBar={() => {}}
        openApp={() => {}}
        ref={ref}
      />
    );

    const winEl = document.getElementById('test-window')!;
    winEl.getBoundingClientRect = () => ({
      left: 1820,
      top: 200,
      right: 1920,
      bottom: 300,
      width: 100,
      height: 100,
      x: 1820,
      y: 200,
      toJSON: () => {}
    });

    act(() => {
      ref.current!.handleDrag();
    });
    act(() => {
      ref.current!.handleStop();
    });

    expect(ref.current!.state.snapped).toBe('right');
    expect(ref.current!.state.width).toBeCloseTo(50);
    const expectedHeight = ((window.innerHeight - 28) / window.innerHeight) * 100;
    expect(ref.current!.state.height).toBeCloseTo(expectedHeight, 5);
    expect(winEl.style.transform).toBe(`translate(${window.innerWidth / 2}px, 0px)`);
  });

  it('snaps window on drag stop near top edge', () => {
    setViewport(1366, 768);
    const ref = React.createRef<Window>();
    render(
      <Window
        id="test-window"
        title="Test"
        screen={() => <div>content</div>}
        focus={() => {}}
        hasMinimised={() => {}}
        closed={() => {}}
        hideSideBar={() => {}}
        openApp={() => {}}
        ref={ref}
      />
    );

    const winEl = document.getElementById('test-window')!;
    winEl.getBoundingClientRect = () => ({
      left: 400,
      top: 6,
      right: 500,
      bottom: 106,
      width: 100,
      height: 100,
      x: 400,
      y: 6,
      toJSON: () => {}
    });

    act(() => {
      ref.current!.handleDrag();
    });
    act(() => {
      ref.current!.handleStop();
    });

    expect(ref.current!.state.snapped).toBe('top');
    expect(ref.current!.state.width).toBeCloseTo(100, 2);
    expect(ref.current!.state.height).toBeCloseTo(50, 2);
    expect(winEl.style.transform).toBe('translate(0px, 0px)');
  });

  it('releases snap with Alt+ArrowDown restoring size', () => {
    setViewport(1024, 768);
    const ref = React.createRef<Window>();
    render(
      <Window
        id="test-window"
        title="Test"
        screen={() => <div>content</div>}
        focus={() => {}}
        hasMinimised={() => {}}
        closed={() => {}}
        hideSideBar={() => {}}
        openApp={() => {}}
        ref={ref}
      />
    );

    const winEl = document.getElementById('test-window')!;
    winEl.getBoundingClientRect = () => ({
      left: 5,
      top: 120,
      right: 105,
      bottom: 220,
      width: 100,
      height: 100,
      x: 5,
      y: 120,
      toJSON: () => {}
    });

    act(() => {
      ref.current!.handleDrag();
    });
    act(() => {
      ref.current!.handleStop();
    });

    expect(ref.current!.state.snapped).toBe('left');
    const snappedHeight = ((window.innerHeight - 28) / window.innerHeight) * 100;
    expect(ref.current!.state.height).toBeCloseTo(snappedHeight, 5);

    act(() => {
      ref.current!.handleKeyDown({
        key: 'ArrowDown',
        altKey: true,
        preventDefault: jest.fn(),
        stopPropagation: jest.fn()
      } as any);
    });

    expect(ref.current!.state.snapped).toBeNull();
    expect(ref.current!.state.width).toBe(60);
    expect(ref.current!.state.height).toBe(85);
  });

  it('releases snap when starting drag', () => {
    setViewport(1440, 900);
    const ref = React.createRef<Window>();
    render(
      <Window
        id="test-window"
        title="Test"
        screen={() => <div>content</div>}
        focus={() => {}}
        hasMinimised={() => {}}
        closed={() => {}}
        hideSideBar={() => {}}
        openApp={() => {}}
        ref={ref}
      />
    );

    const winEl = document.getElementById('test-window')!;
    winEl.getBoundingClientRect = () => ({
      left: 5,
      top: 150,
      right: 105,
      bottom: 250,
      width: 100,
      height: 100,
      x: 5,
      y: 150,
      toJSON: () => {}
    });

    act(() => {
      ref.current!.handleDrag();
    });
    act(() => {
      ref.current!.handleStop();
    });

    expect(ref.current!.state.snapped).toBe('left');
    const snappedHeight = ((window.innerHeight - 28) / window.innerHeight) * 100;
    expect(ref.current!.state.height).toBeCloseTo(snappedHeight, 5);

    act(() => {
      ref.current!.changeCursorToMove();
    });

    expect(ref.current!.state.snapped).toBeNull();
    expect(ref.current!.state.width).toBe(60);
    expect(ref.current!.state.height).toBe(85);
  });
});

describe('Window keyboard dragging', () => {
  it('moves window using arrow keys with grabbed state', () => {
    render(
      <Window
        id="test-window"
        title="Test"
        screen={() => <div>content</div>}
        focus={() => {}}
        hasMinimised={() => {}}
        closed={() => {}}
        hideSideBar={() => {}}
        openApp={() => {}}
      />
    );

    const handle = screen.getByText('Test').parentElement!;

    fireEvent.keyDown(handle, { key: ' ', code: 'Space' });
    fireEvent.keyDown(handle, { key: 'ArrowRight' });

    const winEl = document.getElementById('test-window')!;
    expect(winEl.style.transform).toBe('translate(10px, 0px)');
    expect(handle).toHaveAttribute('aria-grabbed', 'true');

    fireEvent.keyDown(handle, { key: ' ', code: 'Space' });
    expect(handle).toHaveAttribute('aria-grabbed', 'false');
  });
});

describe('Edge resistance', () => {
  it('clamps drag movement near boundaries', () => {
    const ref = React.createRef<Window>();
    render(
      <Window
        id="test-window"
        title="Test"
        screen={() => <div>content</div>}
        focus={() => {}}
        hasMinimised={() => {}}
        closed={() => {}}
        hideSideBar={() => {}}
        openApp={() => {}}
        ref={ref}
      />
    );

    const winEl = document.getElementById('test-window')!;
    winEl.getBoundingClientRect = () => ({
      left: 0,
      top: 0,
      right: 100,
      bottom: 100,
      width: 100,
      height: 100,
      x: 0,
      y: 0,
      toJSON: () => {}
    });

    act(() => {
      ref.current!.handleDrag({}, { node: winEl, x: -100, y: -50 } as any);
    });

    expect(winEl.style.transform).toBe('translate(0px, 0px)');
  });
});

describe('Window overlay inert behaviour', () => {
  it('sets and removes inert on default __next root restoring focus', () => {
    const ref = React.createRef<Window>();
    const root = document.createElement('div');
    root.id = '__next';
    document.body.appendChild(root);

    const opener = document.createElement('button');
    document.body.appendChild(opener);
    opener.focus();

    render(
      <Window
        id="test-window"
        title="Test"
        screen={() => <div>content</div>}
        focus={() => {}}
        hasMinimised={() => {}}
        closed={() => {}}
        hideSideBar={() => {}}
        openApp={() => {}}
        ref={ref}
      />,
      { container: root }
    );

    act(() => {
      ref.current!.activateOverlay();
    });

    expect(root).toHaveAttribute('inert');

    act(() => {
      ref.current!.closeWindow();
    });

    expect(root).not.toHaveAttribute('inert');

    document.body.removeChild(root);
    document.body.removeChild(opener);
  });

  it('respects overlayRoot prop when provided', () => {
    const ref = React.createRef<Window>();
    const root = document.createElement('div');
    root.id = 'custom-root';
    document.body.appendChild(root);

    const opener = document.createElement('button');
    document.body.appendChild(opener);
    opener.focus();

    render(
      <Window
        id="test-window"
        title="Test"
        screen={() => <div>content</div>}
        focus={() => {}}
        hasMinimised={() => {}}
        closed={() => {}}
        hideSideBar={() => {}}
        openApp={() => {}}
        ref={ref}
        overlayRoot="custom-root"
      />,
      { container: root }
    );

    act(() => {
      ref.current!.activateOverlay();
    });

    expect(root).toHaveAttribute('inert');

    act(() => {
      ref.current!.closeWindow();
    });

    expect(root).not.toHaveAttribute('inert');

    document.body.removeChild(root);
    document.body.removeChild(opener);
  });
});
