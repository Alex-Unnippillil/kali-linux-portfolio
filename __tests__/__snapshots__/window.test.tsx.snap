// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Window focus affordance annotates focus state for styling hooks 1`] = `
<DocumentFragment>
  <div>
    <div
      aria-label="Focus Test"
      class="cursor-default rounded-lg rounded-b-none z-30 window--focused opened-window overflow-hidden min-w-1/4 min-h-1/4 main-window absolute window-shadow border-black border-opacity-40 border border-t-0 flex flex-col"
      data-focused="true"
      id="focus-window"
      role="dialog"
      style="width: 60%; height: 85%;"
      tabindex="0"
    >
      <div
        class="windowYBorder cursor-[e-resize] border-transparent border-1 absolute top-1/2 left-1/2 transform -translate-x-1/2 -translate-y-1/2"
      />
      <div
        class="windowXBorder cursor-[n-resize] border-transparent border-1 absolute top-1/2 left-1/2 transform -translate-x-1/2 -translate-y-1/2"
      />
      <div
        aria-grabbed="false"
        class=" relative bg-ub-window-title border-t-2 border-white border-opacity-5 px-3 text-white w-full select-none rounded-b-none flex items-center h-11"
        role="button"
        tabindex="0"
      >
        <div
          class="flex justify-center w-full text-sm font-bold"
        >
          Focus Test
        </div>
      </div>
      <div
        class="absolute select-none right-0 top-0 mt-1 mr-1 flex justify-center items-center h-11 min-w-[8.25rem]"
      >
        <button
          aria-label="Window minimize"
          class="mx-1 bg-white bg-opacity-0 hover:bg-opacity-10 rounded-full flex justify-center items-center h-6 w-6"
          type="button"
        >
          <img
            alt="Kali window minimize"
            class="h-4 w-4 inline"
            data-nimg="1"
            decoding="async"
            height="16"
            loading="lazy"
            src="/themes/Yaru/window/window-minimize-symbolic.svg"
            style="color: transparent;"
            width="16"
          />
        </button>
        <button
          aria-label="Window maximize"
          class="mx-1 bg-white bg-opacity-0 hover:bg-opacity-10 rounded-full flex justify-center items-center h-6 w-6"
          type="button"
        >
          <img
            alt="Kali window maximize"
            class="h-4 w-4 inline"
            data-nimg="1"
            decoding="async"
            height="16"
            loading="lazy"
            src="/themes/Yaru/window/window-maximize-symbolic.svg"
            style="color: transparent;"
            width="16"
          />
        </button>
        <button
          aria-label="Window close"
          class="mx-1 focus:outline-none cursor-default bg-ub-cool-grey bg-opacity-90 hover:bg-opacity-100 rounded-full flex justify-center items-center h-6 w-6"
          id="close-focus-window"
          type="button"
        >
          <img
            alt="Kali window close"
            class="h-4 w-4 inline"
            data-nimg="1"
            decoding="async"
            height="16"
            loading="lazy"
            src="/themes/Yaru/window/window-close-symbolic.svg"
            style="color: transparent;"
            width="16"
          />
        </button>
      </div>
      <div
        class="w-full flex-grow z-20 max-h-full overflow-y-auto windowMainScreen bg-ub-cool-grey"
      >
        <div>
          content
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
